pkg/ssl.go

```diff
@@ -0,0 +1,188 @@
+		privateKeyFile = tempDerFile.Name()
+	}
+
+	lock := s.lock(keyStorePassword, trustStorePassword, certificateFile, privateKeyFile, httpsHostname, httpsPort)
+	check, err := lock.State()
+	if err != nil {
+		return false, err
+	}
+	if check.UpToDate {
+		log.Debugf("%s > SSL already set up (up-to-date)", s.instance.ID())
+		return false, nil
+	}
+
+	params := map[string]any{
+		"keystorePassword":          keyStorePassword,
+		"keystorePasswordConfirm":   keyStorePassword,
+		"truststorePassword":        trustStorePassword,
+		"truststorePasswordConfirm": trustStorePassword,
+		"httpsHostname":             httpsHostname,
+		"httpsPort":                 httpsPort,
+	}
+
+	response, err := s.instance.http.
+		RequestFormData(params).
+		SetFiles(map[string]string{
+			"certificateFile": certificateFile,
+			"privatekeyFile":  privateKeyFile,
+		}).
+		SetDoNotParseResponse(true).
+		Post(SSLSetupPath)
+
+	if err != nil {
+		return false, fmt.Errorf("%s > failed to setup SSL: %w", s.instance.ID(), err)
+	} else if response.IsError() {
+		rawBody := response.RawBody()
+		if rawBody == nil {
+			return false, fmt.Errorf("%s > failed to setup SSL: %s", s.instance.ID(), response.Status())
+		}
+		defer rawBody.Close()
+		body, err := io.ReadAll(rawBody)
+		if err != nil {
+			return false, fmt.Errorf("%s > failed to setup SSL: %s, %w", s.instance.ID(), response.Status(), err)
+		}
+		errorMessage, err := s.getErrorMessage(string(body[:]))
+		if err != nil {
+			return false, fmt.Errorf("%s > failed to setup SSL: %s, %w", s.instance.ID(), response.Status(), err)
+		}
+		return false, fmt.Errorf("%s > failed to setup SSL: %s, %s", s.instance.ID(), response.Status(), errorMessage)
+	}
+
+	if err := lock.Lock(); err != nil {
+		return true, fmt.Errorf("%s > failed to lock SSL setup: %w", s.instance.ID(), err)
+	}
+
+	return true, nil
+}
+
+func (s SSL) writeDER(tempDerFile *os.File, pemBlock *pem.Block) error {
+	if _, err := tempDerFile.Write(pemBlock.Bytes); err != nil {
+		return err
+	}
+	err := tempDerFile.Close()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// From HTML response body, e.g.:
+// <!DOCTYPE html>
+// <html lang='en'>
+// <head>
+// <title>Error</title>
+// </head>
+// <body>
+// <h1>Error</h1>
+// <dl>
+// <dt class='foundation-form-response-status-code'>Status</dt>
+// <dd>500</dd>
+// <dt class='foundation-form-response-status-message'>Message</dt>
+// <dd>Invalid password for existing key store</dd>
+// <dt class='foundation-form-response-title'>Title</dt>
+// <dd>Error</dd>
+// </dl>
+// </body>
+// </html>
+// returns "Invalid password for existing key store"
+func (s SSL) getErrorMessage(body string) (errorMessage string, err error) {
+	lines := strings.Split(string(body), "\n")
+	errorMessageAhead := false
+	for _, line := range lines {
+		if strings.Contains(line, "foundation-form-response-status-message") {
+			errorMessageAhead = true
+			continue
+		}
+		if errorMessageAhead {
+			line = strings.Split(line, ">")[1]
```

Comment:
store the result of split to var; then reuse

Reply:
nvm, you are right

Reply:
I'm not reusing the value o split. I added some comments to make it clear